Sync committee message validation exception

This is happening when running local interop with 2 nodes, after #7171 
```
2023-05-26 08:52:33.354 WARN  - Encountered exception while processing message for topic /eth2/962f1553/sync_committee_3/ssz_snappy
java.util.concurrent.CompletionException: java.util.ConcurrentModificationException
at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368) ~[?:?]
at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377) ~[?:?]
at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:146) ~[classes/:?]
at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887) ~[?:?]
at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325) ~[?:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:626) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:33) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:143) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateTo(SafeFuture.java:320) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:124) ~[classes/:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.util.ConcurrentModificationException
at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:756) ~[?:?]
at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:778) ~[?:?]
at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1856) ~[?:?]
at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129) ~[?:?]
at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527) ~[?:?]
at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513) ~[?:?]
at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150) ~[?:?]
at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647) ~[?:?]
at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessageValidator.validate(SyncCommitteeMessageValidator.java:120) ~[classes/:?]
at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.add(SyncCommitteeMessagePool.java:78) ~[classes/:?]
at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.addRemote(SyncCommitteeMessagePool.java:72) ~[classes/:?]
at tech.pegasys.teku.networking.eth2.gossip.subnets.SyncCommitteeSubnetSubscriptions.lambda$createTopicHandler$0(SyncCommitteeSubnetSubscriptions.java:73) ~[classes/:?]
at tech.pegasys.teku.networking.eth2.gossip.topics.topichandlers.Eth2TopicHandler.lambda$handleMessage$2(Eth2TopicHandler.java:113) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:74) ~[classes/:?]
... 4 more
2023-05-26 08:52:33.355 WARN  - Encountered exception while processing message for topic /eth2/962f1553/sync_committee_0/ssz_snappy
java.util.concurrent.CompletionException: java.util.ConcurrentModificationException
at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368) ~[?:?]
at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377) ~[?:?]
at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:146) ~[classes/:?]
at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887) ~[?:?]
at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325) ~[?:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:626) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:33) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:143) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateTo(SafeFuture.java:320) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:124) ~[classes/:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.util.ConcurrentModificationException
at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:756) ~[?:?]
at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:778) ~[?:?]
at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1856) ~[?:?]
at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129) ~[?:?]
at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527) ~[?:?]
at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513) ~[?:?]
at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150) ~[?:?]
at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647) ~[?:?]
at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessageValidator.validate(SyncCommitteeMessageValidator.java:120) ~[classes/:?]
at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.add(SyncCommitteeMessagePool.java:78) ~[classes/:?]
at tech.pegasys.teku.statetransition.synccommittee.SyncCommitteeMessagePool.addRemote(SyncCommitteeMessagePool.java:72) ~[classes/:?]
at tech.pegasys.teku.networking.eth2.gossip.subnets.SyncCommitteeSubnetSubscriptions.lambda$createTopicHandler$0(SyncCommitteeSubnetSubscriptions.java:73) ~[classes/:?]
at tech.pegasys.teku.networking.eth2.gossip.topics.topichandlers.Eth2TopicHandler.lambda$handleMessage$2(Eth2TopicHandler.java:113) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:74) ~[classes/:?]
... 4 more
```

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessageValidator.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessageValidator.java
@@ -114,23 +114,28 @@ public class SyncCommitteeMessageValidator {
       final Optional<Bytes32> maybeBestBlockRoot = recentChainData.getBestBlockRoot();
 
       if (maybeBestBlockRoot.isPresent()) {
-        final Optional<Bytes32> bestSeenRoot =
-            seenIndices.stream()
-                .filter(item -> item.isSameIgnoringBlockRoot(key))
-                .findFirst()
-                .map(UniquenessKey::getBlockRoot);
+        final Optional<Bytes32> bestSeenRoot;
+        synchronized (this) {
+          bestSeenRoot =
+              seenIndices.stream()
+                  .filter(item -> item.isSameIgnoringBlockRoot(key))
+                  .findFirst()
+                  .map(UniquenessKey::getBlockRoot);
+        }
         // I've already seen this message, can ignore it.
         if (bestSeenRoot.isPresent() && maybeBestBlockRoot.get().equals(bestSeenRoot.get())) {
           return completedFuture(IGNORE);
         } else {
-          if (seenIndices.remove(key)) {
-            // I've seen a message for this slot already, and its block root didn't match current
-            // head
-            LOG.trace(
-                "Removed already seen sync committee message from cache "
-                    + "to accept a better one for validator index {}, slot {}",
-                message.getValidatorIndex(),
-                message.getSlot());
+          synchronized (this) {
+            if (seenIndices.remove(key)) {
+              // I've seen a message for this slot already, and its block root didn't match current
+              // head
+              LOG.trace(
+                  "Removed already seen sync committee message from cache "
+                      + "to accept a better one for validator index {}, slot {}",
+                  message.getValidatorIndex(),
+                  message.getSlot());
+            }
           }
         }
       }
@@ -193,8 +198,10 @@ public class SyncCommitteeMessageValidator {
 
     // [IGNORE] There has been no other valid sync committee message for the declared slot for the
     // validator referenced by sync_committee_message.validator_index.
-    if (seenIndices.containsAll(uniquenessKeys)) {
-      return completedFuture(IGNORE);
+    synchronized (this) {
+      if (seenIndices.containsAll(uniquenessKeys)) {
+        return completedFuture(IGNORE);
+      }
     }
 
     // [REJECT] The subnet_id is correct, i.e. subnet_id in
@@ -228,9 +235,11 @@ public class SyncCommitteeMessageValidator {
               if (!signatureValid) {
                 return reject("Rejecting sync committee message because the signature is invalid");
               }
-              if (!seenIndices.addAll(uniquenessKeys)) {
-                return ignore(
-                    "Ignoring sync committee message as a duplicate was processed during validation");
+              synchronized (this) {
+                if (!seenIndices.addAll(uniquenessKeys)) {
+                  return ignore(
+                      "Ignoring sync committee message as a duplicate was processed during validation");
+                }
               }
               return ACCEPT;
             });

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessageValidator.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeMessageValidator.java
@@ -47,7 +47,7 @@ import tech.pegasys.teku.storage.client.RecentChainData;
 public class SyncCommitteeMessageValidator {
   private static final Logger LOG = LogManager.getLogger();
   private final Set<UniquenessKey> seenIndices =
-      LimitedSet.createSynchronized(VALID_SYNC_COMMITTEE_MESSAGE_SET_SIZE * 2);
+      LimitedSet.createIterable(VALID_SYNC_COMMITTEE_MESSAGE_SET_SIZE * 2);
   private final Spec spec;
   private final SyncCommitteeStateUtils syncCommitteeStateUtils;
   private final AsyncBLSSignatureVerifier signatureVerifier;
@@ -114,28 +114,23 @@ public class SyncCommitteeMessageValidator {
       final Optional<Bytes32> maybeBestBlockRoot = recentChainData.getBestBlockRoot();
 
       if (maybeBestBlockRoot.isPresent()) {
-        final Optional<Bytes32> bestSeenRoot;
-        synchronized (this) {
-          bestSeenRoot =
-              seenIndices.stream()
-                  .filter(item -> item.isSameIgnoringBlockRoot(key))
-                  .findFirst()
-                  .map(UniquenessKey::getBlockRoot);
-        }
+        final Optional<Bytes32> bestSeenRoot =
+            seenIndices.stream()
+                .filter(item -> item.isSameIgnoringBlockRoot(key))
+                .findFirst()
+                .map(UniquenessKey::getBlockRoot);
         // I've already seen this message, can ignore it.
         if (bestSeenRoot.isPresent() && maybeBestBlockRoot.get().equals(bestSeenRoot.get())) {
           return completedFuture(IGNORE);
         } else {
-          synchronized (this) {
-            if (seenIndices.remove(key)) {
-              // I've seen a message for this slot already, and its block root didn't match current
-              // head
-              LOG.trace(
-                  "Removed already seen sync committee message from cache "
-                      + "to accept a better one for validator index {}, slot {}",
-                  message.getValidatorIndex(),
-                  message.getSlot());
-            }
+          if (seenIndices.remove(key)) {
+            // I've seen a message for this slot already, and its block root didn't match current
+            // head
+            LOG.trace(
+                "Removed already seen sync committee message from cache "
+                    + "to accept a better one for validator index {}, slot {}",
+                message.getValidatorIndex(),
+                message.getSlot());
           }
         }
       }
@@ -198,10 +193,8 @@ public class SyncCommitteeMessageValidator {
 
     // [IGNORE] There has been no other valid sync committee message for the declared slot for the
     // validator referenced by sync_committee_message.validator_index.
-    synchronized (this) {
-      if (seenIndices.containsAll(uniquenessKeys)) {
-        return completedFuture(IGNORE);
-      }
+    if (seenIndices.containsAll(uniquenessKeys)) {
+      return completedFuture(IGNORE);
     }
 
     // [REJECT] The subnet_id is correct, i.e. subnet_id in
@@ -235,11 +228,9 @@ public class SyncCommitteeMessageValidator {
               if (!signatureValid) {
                 return reject("Rejecting sync committee message because the signature is invalid");
               }
-              synchronized (this) {
-                if (!seenIndices.addAll(uniquenessKeys)) {
-                  return ignore(
-                      "Ignoring sync committee message as a duplicate was processed during validation");
-                }
+              if (!seenIndices.addAll(uniquenessKeys)) {
+                return ignore(
+                    "Ignoring sync committee message as a duplicate was processed during validation");
               }
               return ACCEPT;
             });

